lib/msf/core/post/windows/mssql.rb

```diff
@@ -0,0 +1,165 @@
+# -*- coding: binary -*-
+
+module Msf
+class Post
+module Windows
+
+module MSSQL
+
+  attr_accessor :sql_client
+
+  include Msf::Exploit::Remote::MSSQL_COMMANDS
+  include Msf::Post::Windows::Services
+  include Msf::Post::Windows::Priv
+
+  def check_for_sqlserver(instance=nil)
+    target_service = nil
+    each_service do |service|
+      unless instance.to_s.strip.empty?
+        if service[:display].downcase.include?(instance.downcase)
+          target_service = service
+          break
+        end
+      else
+        # Target default instance
+        if service[:display] =~ /SQL Server \(| MSSQLSERVER/i
+          target_service = service
+          break
+        end
+      end
+    end
+
+    if target_service
+      target_service.merge!(service_info(target_service[:name]))
+    end
+
+    return target_service
+  end
+
+  def get_sql_client
+    client = nil
+
+    if check_sqlcmd
+      client = 'sqlcmd'
+    elsif check_osql
+      client = 'osql'
+    end
+
+    @sql_client = client
+    return client
+  end
+
+  def check_osql
+    running_services1 = run_cmd("osql -?")
+    services_array1 = running_services1.split("\n")
+    return services_array1.join =~ /(SQL Server Command Line Tool)|(usage: osql)/
+  end
+
+  def check_sqlcmd
+    running_services = run_cmd("sqlcmd -?")
+    services_array = running_services.split("\n")
+    services_array.each do |service|
+      if service =~ /SQL Server Command Line Tool/
+        return true
+      end
+    end
+  end
+
+  def run_sql(query, instance=nil, server='.')
+    target = server
+    if instance && instance.downcase != 'mssqlserver'
+      target = "#{server}\\#{instance}"
+    end
+    cmd = "#{@sql_client} -E -S #{target} -Q \"#{query}\" -h-1 -w 200"
+    vprint_status(cmd)
+    run_cmd(cmd)
+  end
+
+  ## ----------------------------------------------
+  ## Method for executing cmd and returning the response
+  ##
+  ## Note: This is from one of Jabra's modules - Thanks man!
+  ## Note: This craps out when escalating from local admin to system
```

Comment:
Normal processes won't have Impersonate privileges, even when escalated to SYSTEM. User will need to migrate into an authentic SYSTEM process to obtain correct SeImpersonate/SeTcb/SeAssignPrimaryToken privileges for impersonation to work...